src/Psalm/Internal/Type/Comparator/UnionTypeComparator.php

```diff
@@ -48,8 +48,11 @@ public static function isContainedBy(
             return true;
         }
 
-        $container_has_template = $container_type->hasTemplateOrStatic();
+        if ($input_type->isFormerStaticObject()) {
```

Comment:
Can you explain what that does? I find it weird to be able to assert a type is contained by the other just by looking at one of the types

Reply:
When calculating, static types are now resolved to their class with the `was_static` property true

```php
class A {
    public foo(): static {}
}

class B extends A {}

$b = new B;
$foo = $b->foo(); // $foo == "A&static"

assert($foo instanceof B); // Should work
```
In this situation, `isContainedBy` is called to test that `B` is contained by `$foo` but to psalm its `A&static`, so this will see if it `was_static` and return true because `B::foo` dose contain `A::foo` if the return type of `A::foo` is `static`

Hope that makes sense, its really had to explain in a comment lol.

Reply:
when I take your example and run it against your code, there's no difference whether or not the call to isFormerStaticObject is there. Do you have an example of code that is resolved by this method call? I think we need to include an additional check with $container_type to the mix, but I don't see what is the issue being resolved here

Reply:
This was from when I was fixing the test I added in tests/MethodSignatureTest.php:613

Reply:
Likewise, I have no errors with or without this condition.

Can you remove it from your PR so we see the impact on every test?

Reply:
That has now been removed :+1:

Reply:
Ok, this snippet: https://psalm.dev/r/fcd88c33ed (or MethodSignatureTest with returnStaticParent) fails without your piece of code.

Basically, isContainedBy receives an $input_type `A&static` and a $container_type `B&static`. Those are indeed compatible, but you can't know that by just looking at if `A&static` contains `static`, you have to check that A is a parent of B

Reply:
OK so now I am really stuck. What I have so far is.

```php
if ($input_type->equals($container_type)) {
    return true;
}

if ($input_type->isFormerStaticObject()
 && $codebase->classExtendsOrImplements((string)$container_type, (string)$input_type)
) {
    return true;
}
```

The issue I am having is with the interface test.

```php
abstract class A {
    public function foo(): static {
        return $this;
    }
}

interface I {
    /** @return I */
    public function foo();
}

class C extends A implements I {}
```

When calling `isContainedBy` I am getting `A` for `input_type` and `I` for `container_type`. `A` itself dose not implement `I` but we are checking `C` that dose. In `isContainedBy` I have no access to `C` so cant calculate this, dose this need to be calculated in the function calling `isContainedBy`

Reply:
The issue is that this snippet was broken even before your PR: https://psalm.dev/r/2c9b8b1f70

So, trying to have the same behaviour between phpdoc (which is buggy) to another buggy solution is likely to dig up nasty things.

Honestly, I'm quite lost myself here. I'm pretty sure we should be able to find a method that resolve static usages to the actual called class by using some context, but we'll likely hit another wall there.

The solution is less obvious than I thought it would be (unfortunately, it's quite often with Psalm!), so I'd suggest changing strategy and finding where is the runaway memory consumption and try to stop it from happening. It will probably emit strange errors (or no errors when it should) but at least, Psalm will be able to run correctly